From b3f522ac7e563bf4e2560a8a55e79a7577ed2148 Mon Sep 17 00:00:00 2001
From: jkar8572 <jkar8572>
Date: Mon, 14 Dec 2009 22:27:33 +0000
Subject: [PATCH] Convert quota limits using rq_bsize from RPC request

Back-ported upstream commit aee64e6dcd0366cd826587361b0becdf90a33608,
Changelog and common.h updates removed.
---
 rquota_client.c |   20 ++------------------
 rquota_server.c |    4 ++--
 2 files changed, 4 insertions(+), 20 deletions(-)

diff --git a/rquota_client.c b/rquota_client.c
index 883df40..5de3d2a 100644
--- a/rquota_client.c
+++ b/rquota_client.c
@@ -45,8 +45,8 @@ static inline void clinet2utildqblk(struct util_dqblk *u, struct rquota *n)
 	time_t now;
 	
 	/* Copy the quota */
-	u->dqb_bhardlimit = n->rq_bhardlimit;
-	u->dqb_bsoftlimit = n->rq_bsoftlimit;
+	u->dqb_bhardlimit = toqb(((qsize_t)n->rq_bhardlimit) * n->rq_bsize);
+	u->dqb_bsoftlimit = toqb(((qsize_t)n->rq_bsoftlimit) * n->rq_bsize);
 	u->dqb_ihardlimit = n->rq_fhardlimit;
 	u->dqb_isoftlimit = n->rq_fsoftlimit;
 	u->dqb_curinodes = n->rq_curfiles;
@@ -60,22 +60,6 @@ static inline void clinet2utildqblk(struct util_dqblk *u, struct rquota *n)
 		u->dqb_itime = n->rq_ftimeleft + now;
 	else
 		u->dqb_itime = 0;
-	/* Convert from remote block size */
-	if (n->rq_bsize != RPC_DQBLK_SIZE) {
-		int conversion_unit;
-
-		conversion_unit = n->rq_bsize >> RPC_DQBLK_SIZE_BITS;
-		if (conversion_unit == 0) {
-			conversion_unit = RPC_DQBLK_SIZE / n->rq_bsize;
-
-			u->dqb_bhardlimit /= conversion_unit;
-			u->dqb_bsoftlimit /= conversion_unit;
-		}
-		else {
-			u->dqb_bhardlimit *= conversion_unit;
-			u->dqb_bsoftlimit *= conversion_unit;
-		}
-	}
 }
 
 /* Convert utils format of quotas to network one */
diff --git a/rquota_server.c b/rquota_server.c
index c610636..44f95ad 100644
--- a/rquota_server.c
+++ b/rquota_server.c
@@ -95,8 +95,8 @@ static inline void servutil2netdqblk(struct rquota *n, struct util_dqblk *u)
 	time_t now;
 
 	time(&now);
-	n->rq_bhardlimit = u->dqb_bhardlimit;
-	n->rq_bsoftlimit = u->dqb_bsoftlimit;
+	n->rq_bhardlimit = (u->dqb_bhardlimit << QUOTABLOCK_BITS) >> RPC_DQBLK_SIZE_BITS;
+	n->rq_bsoftlimit = (u->dqb_bsoftlimit << QUOTABLOCK_BITS) >> RPC_DQBLK_SIZE_BITS;
 	n->rq_fhardlimit = u->dqb_ihardlimit;
 	n->rq_fsoftlimit = u->dqb_isoftlimit;
 	n->rq_curblocks = (u->dqb_curspace + RPC_DQBLK_SIZE - 1) >> RPC_DQBLK_SIZE_BITS;
-- 
1.7.3.4

